examples/vae.py

```diff
@@ -63,8 +63,8 @@ def binarize(rng, batch):
 def main(args):
     encoder_init, encode = encoder(args.hidden_dim, args.z_dim)
     decoder_init, decode = decoder(args.hidden_dim, 28 * 28)
-    opt_init, opt_update, get_params = optimizers.adam(args.learning_rate)
-    svi_init, svi_update, svi_eval = svi(model, guide, elbo, opt_init, opt_update, get_params,
+    opt_init, opt_update, get_opt_params = optimizers.adam(args.learning_rate)
```

Comment:
We need to initialize the optimizer outside the svi function so that we can pass `get_params` to autoguides.

Reply:
How about creating an arg to get_params method to achieve that purpose?

Reply:
Hmm..I'm not sure if that would help, specially if `get_params` is an internal function of `svi`, since we have to declare the autoguide before `svi`. Or maybe I'm misunderstanding your suggestion?

Reply:
My idea is to make AutoGuide separated from `opt_state`. We just need to provide `opt_get_params(opt_state)` to AutoGuide methods to obtain the results. So I think we can achieve it by `params = svi_get_params(opt_state, unconstrained=True)` and use this `params` in autoguide methods.

Reply:
Hopefully, this shouldn't matter if we convert it into a class. Its fine to only send constrained parameters to autoguide methods directly, but then the onus of getting the constrained parameters falls on the user.

Reply:
>  the onus of getting the constrained parameters falls on the user

I want the simplicity too. Assume that we use classes and have `optimizer = optim.Adam(...)`, then users need to either
```
guide = AutoDiagonalNormal(optimizer)
guide.quantile(svi_state)
```
or
```
guide = AutoDiagonalNormal()
guide.quantile(optimizer.get_params(svi_state.opt_state))
```
It seems that the first way (which is more aligned with current interface) is simpler.

Another way is to have an additional field in `svi_state`, says `svi_state.opt_params`, so users only need to provide
```
guide.quantile(svi_state.opt_params)
```
WDYT?

Reply:
Either (1) or (2) looks fine to me right now, though (2) looks better since that frees us to declare the optimizer wherever we can.